File: CNV_detekce_final_v6_CNV2/gesd.m
function [outliers, idx] = gesd(coverageSignal, estimatedOutliers)
% GESD Two-sided Generalized (extreme Studentized deviate) ESD test
% coverageSignal vector of coverage
% estimatedOutliers number of estimated over threshold values

%% COMPUTATION
alpha = 0.05;
R=zeros(1,estimatedOutliers);
Lambda=zeros(1,estimatedOutliers);
outliers=zeros(1,estimatedOutliers);

coverageSignal=double(coverageSignal);

n = length(coverageSignal);
xCopy = coverageSignal;
outliersNumber=0;
for i = 1 : estimatedOutliers
    xMean = mean(coverageSignal);
    xSD = std(coverageSignal);
    
    R(i) = max( abs(coverageSignal - xMean) / xSD );
    
    % compute critical value
    p = 1 - alpha /2 /(n - i + 1);
    t = tinv(p, n - i - 1);
    Lambda(i) = t * (n - i) / sqrt((n - i - 1 + t^2) * (n - i + 1));
    
    if R(i) > Lambda(i)
        outliersNumber = i;
    end
    
    [~, idxMax] = max( abs(coverageSignal - xMean) );
    outliers(i) = coverageSignal(idxMax);
    coverageSignal(idxMax) = [];
end

idx = zeros(1, outliersNumber);

for i = 1 : outliersNumber
    idx(i) = find(xCopy == outliers(i), 1);
    xCopy( idx(i) ) = NaN;
end

if outliersNumber > 0
    outliers = outliers(1 : outliersNumber);
    [idx,order]=sort(idx);
    outliers=outliers(order);
else
    outliers=[];
    idx=[];
end




end

File: R/gesd.R
#' GESD - Two-sided Generalized Extreme Studentized Deviate test
#' This function requires:
#' 1 - a dataframe with 2 columns - 1st column: position/coordinate 2nd column: coverage/read-depth 
#' 2- an upper bound for the suspected number of outliers (an integer)
#' It returns outliers indices as vector
#' See https://www.itl.nist.gov/div898/handbook/eda/section3/eda35h3.htm
#' 
#' @param coverageDF dataframe of coverage, 3 columns CHROM POS COVERAGE
#' @param estimatedOutliers Number of potential outliers from M-Zscore function
#' @return A vector of outliers indices from input
#' @export
#' @noRd
#' 
gesd <- function(coverageDF, estimatedOutliers) {
  # COMPUTATION
  coverageSignal <- coverageDF$COVERAGE
  alpha <- 0.05
  R <- replicate(estimatedOutliers, 0)
  Lambda <-  replicate(estimatedOutliers, 0)
  outliers <-  replicate(estimatedOutliers, 0)
  
  n <- length(coverageSignal)
  xCopy  <- coverageSignal
  outliersNumber <- 0
  
  for (i in 1:estimatedOutliers) {
    xMean <- mean(coverageSignal)
    xSD <- sd(coverageSignal)
    
    R[i] <- max(abs(coverageSignal - xMean) / xSD)
    
    # compute critical value
    p <- 1 - alpha / 2 / (n - i + 1)
    t <- qt(p, n - i - 1)
    Lambda[i] <-
      t * (n - i) / sqrt((n - i - 1 + t ^ 2) * (n - i + 1))
    
    if (R[i] > Lambda[i]) {
      outliersNumber <- i
    }
    
    idxMax <- which.max(abs(coverageSignal - xMean))
    outliers[i] <- coverageSignal[idxMax]
    coverageSignal <-  coverageSignal[-idxMax]
  }
  
  idx <- replicate(outliersNumber, 0)
  
  # match find values in original vector
  if (outliersNumber > 0){
    for (i in 1:outliersNumber) {
      idx[i] <- which(xCopy == outliers[i])[[1]]
      xCopy[idx[i]] <- NaN
      idx <- unlist(idx)
    }
  }
  
  # sort
  if (outliersNumber > 0) {
    coverageDF$isOutlier <- FALSE
    coverageDF$isOutlier[idx] <- TRUE
    return(as.vector(coverageDF$POS[idx]))
  } else {
    return()
    warning("No outliers found")
  }
}
